docs/arcadia-users-group/20220919-aws-s3-cli/lesson.md

```diff
@@ -0,0 +1,301 @@
+
+### ls
+
+In the [first shell workshop](../20220906-intro-to-shell1/lesson.md), we learned that `ls` lists contents of a directory (**l**i**s**t). It serves the same function for the S3 CLI.
+
+Let's start with listing all the Arcadia Science S3 buckets:
+
+```{bash}
+aws s3 ls
+```
+
+Now let's explore what one of these buckets looks like, starting with the `aug-workshop-demo` bucket.
+
+```{bash}
+aws s3 ls s3://aug-workshop-demo
+```
+
+If everything went correctly, you should see the `demo` folder under the `aug-workshop-demo` bucket.
+
+### cp
+
+In the [second shell workshop](../20220912-intro-to-shell2/lesson.md), we learned that `cp` copies a file or directory. The S3 command also serves a similar purpose. Let's try it out by trying to copy the file we created earlied `copy.txt` to the S3 bucket under a folder with your initials. For me, this would be in `s3://aug-workshop-demo/fmc/`.
+
+```{bash}
+aws s3 cp copy.txt s3://aug-workshop-demo/fmc/copy.txt
+```
+
+If you ignore the last part of the S3 path, it'll create a file with the same name:
+
+```{bash}
+aws s3 cp copy.txt s3://aug-workshop-demo/fmc/
+```
+
+Alternatively, you can give a new name to the file on S3:
+
+```{bash}
+aws s3 cp copy.txt s3://aug-workshop-demo/fmc/new-copy.txt
+```
+
+Let's see the changes:
+```{bash}
+ls
+```
+The file is still in our directory, it hasn't moved.
+
+```{bash}
+aws s3 ls s3://aug-workshop-demo
+aws s3 ls s3://aug-workshop-demo/fmc/
+```
+
+Now let's try copying the `sync` directory by:
+```{bash}
+aws s3 cp sync/ s3://aug-workshop-demo/fmc/
+```
+
+That failed! This is because the `cp` command directly works with individual files. For copying folders, we need to use the `--recursive` flag. So, let's try this:
+
+```{bash}
+aws s3 cp --recursive sync/ s3://aug-workshop-demo/fmc/
+aws s3 ls s3://aug-workshop-demo/fmc/
+```
+
+### mv
+
+In the [second shell workshop](../20220912-intro-to-shell2/lesson.md), we learned that `mv` moves or renames a file or directory. The S3 command also serves a similar purpose. Let's start with a moving example by trying to move the `move.txt` file we created earlier to the S3 bucket under a folder with your name. For me, this would be in `s3://aug-workshop-demo/mert/`.
+
+```{bash}
+aws s3 mv move.txt s3://aug-workshop-demo/mert/move.txt
+```
+
+Let's see the changes:
+```{bash}
+ls
+```
+This time, the file is no longer in our directory. Let's check AWS S3:
+
+```{bash}
+aws s3 ls s3://aug-workshop-demo
+aws s3 ls s3://aug-workshop-demo/mert/
+```
+
+Now, let's do a move within S3 by moving the file from `s3://aug-workshop-demo/mert/move.txt` to `s3://aug-workshop-demo/fmc/move.txt`:
+
+```{bash}
+aws s3 mv s3://aug-workshop-demo/mert/move.txt s3://aug-workshop-demo/fmc/move.txt
+```
+
+And now let's visulize it:
+```{bash}
+aws s3 ls s3://aug-workshop-demo/mert/
+aws s3 ls s3://aug-workshop-demo/fmc/
+```
+
+The former folder is empty and the latter now has the `move.txt`.
+
```

Comment:
I'm curious if the mv section -- is there any danger in doing this? Like is it less safe for file integrity than cp, or does AWS do all the necessary checks like md5sums prior to deleting your local copy?

Reply:
Added a note on data integrity but yeah AWS does all the checks: https://docs.aws.amazon.com/cli/latest/topic/s3-faq.html

I still almost never use it though. I am a copy then clean-up person.